Source/Core/Core/IOS/Network/Socket.cpp

```diff
@@ -712,24 +715,82 @@ void WiiSockMan::Update()
+    return;
+  }
 
-  if (ret >= 0)
+  int n = 0;
+  for (auto& elem : WiiSockets)
   {
-    for (auto& pair : WiiSockets)
-    {
-      WiiSocket& sock = pair.second;
-      sock.Update(FD_ISSET(sock.fd, &read_fds) != 0, FD_ISSET(sock.fd, &write_fds) != 0,
-                  FD_ISSET(sock.fd, &except_fds) != 0);
-    }
+    elem.second.Update(!!(ufds[n].revents & read_event), !!(ufds[n].revents & write_event),
+                       !!(ufds[n].revents & error_event));
+    n++;
   }
-  else
+
+  UpdatePollCommands();
+}
+
+void WiiSockMan::UpdatePollCommands()
+{
+  static constexpr int error_event = (POLLHUP | POLLERR);
+
+  if (pending_polls.empty())
+    return;
+
+  const auto now = std::chrono::high_resolution_clock::now();
+  const auto elapsed_d = std::chrono::duration_cast<std::chrono::milliseconds>(now - last_time);
+  const auto elapsed = elapsed_d.count();
+  last_time = now;
+
+  for (auto& pcmd : pending_polls)
   {
-    for (auto& elem : WiiSockets)
-    {
-      elem.second.Update(false, false, false);
-    }
+    // don't touch negative timeouts
+    if (pcmd.timeout > 0)
+      pcmd.timeout = std::max<s64>(0, pcmd.timeout - elapsed);
   }
+
+  pending_polls.erase(
+      std::remove_if(
+          pending_polls.begin(), pending_polls.end(),
+          [this](auto& pcmd) {
+            const int count = std::count_if(pcmd.wii_fds.begin(), pcmd.wii_fds.end(), [&](auto& e) {
+              // Only happens if loaded from a save state
+              if (e.revents & error_event)
+                return true;
+
+              auto poll_it =
+                  find_if(ufds.begin(), ufds.end(), [&](auto& u) { return e.fd == u.fd; });
+
+              // SO_POLL on IOS ignores unknown FDs instead of returning an error
+              if (poll_it == ufds.end())
+                return false;
+
+              return !!((e.events | error_event) & poll_it->revents);
+            });
+
+            if (count == 0 && pcmd.timeout)
+              return false;
+
+            // Translate native to Wii events
+            const auto& pfds = pcmd.wii_fds;
+            for (u32 i = 0; i < pfds.size(); ++i)
+            {
+              const int revents = ConvertEvents(pfds[i].revents, ConvertDirection::NativeToWii);
+
+              // No need to change fd or events as they are input only.
+              // Memory::Write_U32(ufds[i].fd, request.buffer_out + 0xc*i); //fd
+              // Memory::Write_U32(events, request.buffer_out + 0xc*i + 4); //events
+              Memory::Write_U32(revents, pcmd.buffer_out + 0xc * i + 8);  // revents
+              DEBUG_LOG(IOS_NET, "IOCTL_SO_POLL socket %d wevents %08X events %08X revents %08X", i,
+                        revents, pfds[i].events, pfds[i].revents);
+            }
+            GetIOS()->EnqueueIPCReply(Request(pcmd.request_addr), count);
+            return true;
+          }),
+      std::end(pending_polls));
```

Comment:
Is there a reason not to use `pending_polls.end()`? It will get rid of the `iterator` header if I'm not mistaken.